src/nvim/eval/typval_encode.c.h

```diff
@@ -0,0 +1,713 @@
+    typval_T *cur_tv = NULL;
+    switch (cur_mpsv->type) {
+      case kMPConvDict: {
+        if (!cur_mpsv->data.d.todo) {
+          (void)_mp_pop(mpstack);
+          cur_mpsv->data.d.dict->dv_copyID = copyID - 1;
+          TYPVAL_ENCODE_CONV_DICT_END();
+          continue;
+        } else if (cur_mpsv->data.d.todo
+                   != cur_mpsv->data.d.dict->dv_hashtab.ht_used) {
+          TYPVAL_ENCODE_CONV_DICT_BETWEEN_ITEMS();
+        }
+        while (HASHITEM_EMPTY(cur_mpsv->data.d.hi)) {
+          cur_mpsv->data.d.hi++;
+        }
+        dictitem_T *const di = HI2DI(cur_mpsv->data.d.hi);
+        cur_mpsv->data.d.todo--;
+        cur_mpsv->data.d.hi++;
+        TYPVAL_ENCODE_CONV_STR_STRING(&di->di_key[0],
+                                      strlen((char *)&di->di_key[0]));
+        TYPVAL_ENCODE_CONV_DICT_AFTER_KEY();
+        cur_tv = &di->di_tv;
+        break;
+      }
+      case kMPConvList: {
+        if (cur_mpsv->data.l.li == NULL) {
+          (void)_mp_pop(mpstack);
+          cur_mpsv->data.l.list->lv_copyID = copyID - 1;
+          TYPVAL_ENCODE_CONV_LIST_END();
+          continue;
+        } else if (cur_mpsv->data.l.li != cur_mpsv->data.l.list->lv_first) {
+          TYPVAL_ENCODE_CONV_LIST_BETWEEN_ITEMS();
+        }
+        cur_tv = &cur_mpsv->data.l.li->li_tv;
+        cur_mpsv->data.l.li = cur_mpsv->data.l.li->li_next;
+        break;
+      }
+      case kMPConvPairs: {
+        if (cur_mpsv->data.l.li == NULL) {
+          (void)_mp_pop(mpstack);
+          cur_mpsv->data.l.list->lv_copyID = copyID - 1;
+          TYPVAL_ENCODE_CONV_DICT_END();
+          continue;
+        } else if (cur_mpsv->data.l.li != cur_mpsv->data.l.list->lv_first) {
+          TYPVAL_ENCODE_CONV_DICT_BETWEEN_ITEMS();
+        }
+        const list_T *const kv_pair = cur_mpsv->data.l.li->li_tv.vval.v_list;
+        TYPVAL_ENCODE_CONV_SPECIAL_DICT_KEY_CHECK(
+            encode_vim_to__error_ret, kv_pair->lv_first->li_tv);
+        if (_TYPVAL_ENCODE_CONVERT_ONE_VALUE(TYPVAL_ENCODE_FIRST_ARG_NAME,
+                                             &mpstack, cur_mpsv,
+                                             &kv_pair->lv_first->li_tv,
+                                             copyID,
+                                             objname) == FAIL) {
+          goto encode_vim_to__error_ret;
+        }
+        TYPVAL_ENCODE_CONV_DICT_AFTER_KEY();
+        cur_tv = &kv_pair->lv_last->li_tv;
+        cur_mpsv->data.l.li = cur_mpsv->data.l.li->li_next;
+        break;
+      }
+      case kMPConvPartial: {
+        partial_T *const pt = cur_mpsv->data.p.pt;
+        switch (cur_mpsv->data.p.stage) {
+          case kMPConvPartialArgs: {
+            TYPVAL_ENCODE_CONV_FUNC_BEFORE_ARGS(pt == NULL ? 0 : pt->pt_argc);
+            cur_mpsv->data.p.stage = kMPConvPartialSelf;
+            if (pt != NULL && pt->pt_argc > 0) {
+              TYPVAL_ENCODE_CONV_LIST_START(pt->pt_argc);
```

Comment:
In case of **freeing** a partial this does look wrong. For what list is the refcount decreased?

The partial does not have an argument list, but an array of type `typval_T` values. So perhaps instead of using `TYPVAL_ENCODE_CONV_LIST_START()` we could use something like `TYPVAL_ENCODE_CONV_FUNC_LIST_START()`  here that does nothing when freeing and otherwise does what
`TYPVAL_ENCODE_CONV_LIST_START()` does?